File: ruby/Diff.rb

=begin
The algorithm implemented here is based on "An O(NP) Sequence Comparison Algorithm"                                   
by described by Sun Wu, Udi Manber and Gene Myers 
=end

class Diff
  attr_reader :editdis
  def initialize(a, b)
    @a = a
    @b = b
    @m = a.length
    @n = b.length
    if @m > @n
      @a, @b = @b, @a
      @m, @n = @n, @m
    end
    @editdis = 0
  end
  def compose()
    offset = @m + 1
    delta  = @n - @m
    size   = @m + @m + 3
    fp     = Array.new(size){|i| -1}
    p      = -1
    begin
      p = p + 1
      (-p..delta-1).each do |k|
        fp[k+offset] = snake(k, fp[k-1+offset]+1, fp[k+1+offset])
      end
      (delta+p..delta+1).to_a.reverse.each do |k|
        fp[k+offset] = snake(k, fp[k-1+offset]+1, fp[k+1+offset])
      end
      fp[delta+offset] = snake(delta, fp[delta-1+offset]+1, fp[delta+1+offset])
    end while fp[delta+offset] != @n
    @editdis = delta + 2 * p
  end
  def snake(k, p, pp)
    y = [p, pp].max
    x = y - k
    while x < @m and y < @n and @a[x] == @b[y]
      x = x + 1
      y = y + 1
    end
    return y
  end
end

File: objc/diff.m

#import <Foundation/Foundation.h>

#import "diff.h"

/**
 * The algorithm implemented here is based on "An O(NP) Sequence Comparison Algorithm"
 * by described by Sun Wu, Udi Manber and Gene Myers
 */
@implementation Diff
@synthesize editdis;
- (void)swap:(id *)a:(id *)b
{
  id *t;
  t = a;
  a = b;
  b = t;
}

- (void)init:(NSString *)a:(NSString *)b
{
  A = a;
  B = b;
  M = A.length;
  N = B.length;
  if (M > N) {
    [self swap:(id *)A:(id *)B];
    [self swap:(id *)&M:(id *)&N];
  }
}

- (void)compose
{
  NSInteger i, k;
  NSInteger p        = -1;
  NSInteger size     = M + N + 3;
  NSInteger delta    = N - M;
  NSInteger offset   = M + 1;
  NSMutableArray *fp = [[[NSMutableArray alloc] initWithCapacity:size] autorelease];
  NSNumber *op, *opp, *oi, *ly;
  for (i=0;i<size;++i) {
    NSNumber *oi = [NSNumber numberWithInt:-1];
    [fp addObject:oi];
  }
  
  do {
    ++p;
    for (k=-p;k<=delta-1;++k) {
      op  = [fp objectAtIndex:k-1+offset];
      opp = [fp objectAtIndex:k+1+offset];
      oi  = [NSNumber numberWithInt:
                        [self snake:k:[op integerValue]+1:[opp integerValue]]];
      [fp replaceObjectAtIndex:k+offset withObject:oi];
    }
    for (k=delta+p;k>=delta+1;--k) {
      op  = [fp objectAtIndex:k-1+offset];
      opp = [fp objectAtIndex:k+1+offset];
      oi  = [NSNumber numberWithInt:
                        [self snake:k:[op integerValue]+1:[opp integerValue]]];
      [fp replaceObjectAtIndex:k+offset withObject:oi];
    }
    op  = [fp objectAtIndex:delta-1+offset];
    opp = [fp objectAtIndex:delta+1+offset];
    oi  = [NSNumber numberWithInt:
                      [self snake:delta:[op integerValue]+1:[opp integerValue]]];
    [fp replaceObjectAtIndex:delta+offset withObject:oi];
    ly = [fp objectAtIndex:delta+offset];
  } while([ly integerValue] < N);
  editdis = delta + 2 * p;
}

- (NSInteger)snake:(NSInteger)k:(NSInteger)p:(NSInteger)pp
{
  NSInteger y = MAX(p, pp);
  NSInteger x = y - k;
  
  while (x < M && 
         y < N && 
         [A substringWithRange:NSMakeRange(x, 1)] == [B substringWithRange:NSMakeRange(y, 1)])
  {
	++x;
	++y;
  }
  return y;
}
@end
